sandbox/cgroup/monitor.go

```diff
@@ -0,0 +1,194 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"errors"
+	"path"
+	"sync"
+
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/osutil/inotify"
+)
+
+// Manages the inotify watcher, allowing to have a single watch descriptor open
+type inotifyWatcher struct {
+	// The watch object
+	wd *inotify.Watcher
+	// This is used to ensure that the watcher goroutine is launched only once
+	doOnce sync.Once
+	// This channel is used to add a new CGroup that has to be checked
+	addWatchChan chan *groupToWatch
+	// Contains the list of groups to monitor, to detect when they have been deleted
+	groupList []*groupToWatch
+	// Contains the list of paths being monitored by the inotify watcher.
+	// The paths monitored aren't the CGroup paths, but the parent ones, because
+	// in /sys/fs is not possible to detect the InDeleteSelf message, only the
```

Comment:
Is there some documentation or discussions you can share about using inotify in /sys/fs? I thought it wouldn't work since they're it's a pseudo-filesystem and there aren't real files changing (I saw some discussions with Greg Kroah-Hartman online to that effect), so I'm curious to know if that's wrong/outdated info or we're doing something special

Reply:
It is outdated: inotify does work with some parts of pseudo-filesystems, but, as you can see, there are some things that doesn't work, like InDeleteSelf. At the beginning I found this FAQ (https://inotify.aiken.cz/?section=inotify&page=faq&lang=en) where it says:

    Q: Can I watch sysfs (procfs, nfs...)?
    Simply spoken: yes, but with some limitations. These limitations vary between kernel versions and tend to get smaller. Please read information about particular filesystems.

Reply:
Also, in other threads, it says that inotify only notifies things at the filesystem level. That can explain why InDeleteSelf doesn't work, but InDelete does work.

Reply:
Ah cool, thank you for the link